### PR TITLE
JIT: use register indirect for indirect jumps

### LINKED ISSUE #22342: [x64] Test failed: JIT_Methodical._tailcall__il_reldeep_virt__il_reldeep_virt_._tailcall__il_reldeep_virt__il_reldeep_virt_sh
**Job:**
https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/jitstress/job/x64_checked_centos7.1_r2r_jitforcerelocs_tst/171/

**Failed tests:**
`JIT_Methodical._tailcall__il_reldeep_virt__il_reldeep_virt_._tailcall__il_reldeep_virt__il_reldeep_virt_sh`

**Log:**
```
JIT_Methodical._tailcall__il_reldeep_virt__il_reldeep_virt_._tailcall__il_reldeep_virt__il_reldeep_virt_sh [FAIL]
        Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
           at JitTest.TestClass.Method1()
           at JitTest.TestClass.Method0()
           at JitTest.TestClass.Main()
        /mnt/resource/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_centos7.1_r2r_jitforcerelocs_tst/bin/tests/Linux.x64.Checked/JIT/Methodical/tailcall/_il_reldeep_virt/_il_reldeep_virt.sh: line 244: 94280 Aborted                 (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
        /n/nReturn code:      1/nRaw output file:      /mnt/resource/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_centos7.1_r2r_jitforcerelocs_tst/bin/tests/Linux.x64.Checked/JIT/Methodical/Reports/JIT.Methodical/tailcall/_il_reldeep_virt/_il_reldeep_virt.output.txt/nRaw output:/nBEGIN EXECUTION
        in takeLock
        /mnt/resource/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_centos7.1_r2r_jitforcerelocs_tst/bin/tests/Linux.x64.Checked/Tests/Core_Root/crossgen /Platform_Assemblies_Paths /mnt/resource/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_centos7.1_r2r_jitforcerelocs_tst/bin/tests/Linux.x64.Checked/Tests/Core_Root:/mnt/resource/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_centos7.1_r2r_jitforcerelocs_tst/bin/tests/Linux.x64.Checked/JIT/Methodical/tailcall/_il_reldeep_virt/IL:/mnt/resource/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_centos7.1_r2r_jitforcerelocs_tst/bin/tests/Linux.x64.Checked/JIT/Methodical/tailcall/_il_reldeep_virt /in _il_reldeep_virt.org /out _il_reldeep_virt.exe
        Microsoft (R) CoreCLR Native Image Generator - Version 4.5.30319.0
        Copyright (c) Microsoft Corporation.  All rights reserved.
        
        Native image _il_reldeep_virt.exe generated successfully.
        in ReleaseLock
        in takeLock
        /mnt/resource/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_centos7.1_r2r_jitforcerelocs_tst/bin/tests/Linux.x64.Checked/Tests/Core_Root/corerun _il_reldeep_virt.exe
        in ReleaseLock
        Expected: 100
        Actual: 134
        END EXECUTION - FAILED
        Test Harness Exitcode is : 1
        /nTo run the test:/n> set CORE_ROOT=/mnt/resource/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_centos7.1_r2r_jitforcerelocs_tst/bin/tests/Linux.x64.Checked/Tests/Core_Root/n> /mnt/resource/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_centos7.1_r2r_jitforcerelocs_tst/bin/tests/Linux.x64.Checked/JIT/Methodical/tailcall/_il_reldeep_virt/_il_reldeep_virt.sh/n
        Expected: True
        Actual:   False
        Stack Trace:
          /mnt/resource/j/workspace/dotnet_coreclr/master/checked_centos7.1/bin/tests/Linux.x64.Checked/TestWrappers/JIT.Methodical/JIT.Methodical.XUnitWrapper.cs(124849,0): at JIT_Methodical._tailcall__il_reldeep_virt__il_reldeep_virt_._tailcall__il_reldeep_virt__il_reldeep_virt_sh()
```

### PR DESCRIPTION
**Pull request by AndyAyersMS**
If we are jumping to another method via an indirection cell, the cell address
may be too far from the jump to address via a RIP-relative indirection. So
instead, load the target address into RAX and jump register indirect.

Fixes #22342.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

This likely regressed at #21292, as tiered compilation started using indirection cells more pervasively.

**Comment by AndyAyersMS** [comment]
Interestingly there are still a lot of occurrences of `NULL` in codegencommon.cpp. Maybe they're not visible to the formatter legs because of ifdefs?

**Comment by AndyAyersMS** [comment]
Diff from the failing test. Will try and run diffs more broadly but first need to modify `jit-diffs` to pass the "allow tiering" option down to `jit-dasm-pmi`...

```diff
 ; Assembly listing for method JitTest.TestClass:Method1():ref:this
 ; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
...
 G_M63985_IG05:
        488D6500             lea      rsp, [rbp]
        5D                   pop      rbp
-       FF2527F8FFFF         jmp      [JitTest.TestClass:Method2():ref:this]
+       48B8D04BB17925000000 mov      rax, 0x2579B14BD0
+       FF20                 jmp      qword ptr [rax]JitTest.TestClass:Method2():ref:this
```

File: src/jit/codegencommon.cpp [hunk 1]
```diff
@@ -8961,24 +8961,40 @@ void CodeGen::genFnEpilog(BasicBlock* block)
...
+            // If we have IAT_PVALUE, jump via register indirect, as sometimes the
+            // indirection cell can't be reached by the jump.
 
-            // Simply emit a jump to the methodHnd. This is similar to a call so we can use
-            // the same descriptor with some minor adjustments.
+            emitter::EmitCallType callType;
+            void*                 addr;
+            regNumber             indCallReg;
+
+            if (addrInfo.accessType == IAT_PVALUE)
```
Comment (CarolEidt) [inline src/jit/codegencommon.cpp]: I'm not terribly familiar with this code, but this seems to be assuming that `addrInfo.accessType` is either `IAT_VALUE` or `IAT_PVALUE`. Should that be asserted?

**Comment by BruceForstall** [comment]
@AndyAyersMS In the case where the indirection cell doesn't reach, and the VM decides to create a jump stub, isn't that the bug? I.e., it shouldn't be creating a jump stub for a data cell that is not a jump target.

Also, in these cases of data not reaching, shouldn't the entire compilation get aborted, and restarted, with the bit set that we can't create these types of addresses anymore? Thus, in `CEEJitInfo::getRelocTypeHint`, `m_fAllowRel32` will be false.

File: src/jit/codegencommon.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/jit/codegencommon.cpp]: We assert for that at line 8959, just above.

**Comment by AndyAyersMS** [comment]
What I did for `jmp` parallels what we do for indirect calls. For instance earlier in the method there's a call to `Method2` and we emit the `call` the same way that we're now emitting the `jmp`. The call transformations happen in lower but we defer the `jmp` processing until epilog generation.
```asm
       48B8D04BB17925000000 mov      rax, 0x2579B14BD0
       FF10                 call     gword ptr [rax]JitTest.TestClass:Method2():ref:this
       ...
       48B8D04BB17925000000 mov      rax, 0x2579B14BD0
       FF20                 jmp      qword ptr [rax]JitTest.TestClass:Method2():ref:this
```
We don't seem to use `genCodeIndirAddrCanBeEncodedAsPCRelOffset` (which queries the reloc type hint) for indirect method entry points. Perhaps we should -- in some cases we might be able to avoid the 64 bit literals. 

I suspect the presence of `IAT_PVALUE` entry points in jitted code is a relatively new thing (to enable better tier1 backpatching) and perhaps it could use some more polish. In the fragile NGEN case we where we see `IAT_PVALUE` we probably don't have the ability to use 32 bit relocs.

We could have the VM assert if the jit requests a 32 bit reloc to an `IAT_PVALUE` entry point if the VM discovers 32 bits is insufficient, but ultimately it's up to the jit to anticipate this and not ask for such a thing.

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib is Linux arm testing known to be broken?

**Comment by echesakovMSFT** [comment]
> @dotnet/jit-contrib is Linux arm testing known to be broken?

Yes, it is https://github.com/dotnet/core-eng/issues/6686#issuecomment-503630450

**Pull request merged**